mlflow/mlflow
# Add pyenv + virtualenv support for MLflow Models

Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

## What changes are proposed in this pull request?

This PR makes the following changes to support pyenv + virtualenv for `mlflow models`. I'm happy to split the PR to facilitate code review.

- [x] Add `--env-manager` option.
- [x] Deprecate `--no-conda` and encourage using `--env-manager` instead.
- [x] Raise an experimental warning if `virtualenv` is chosen.
- [x] Indicate conda is discouraged now and virtualenv is available as an alternative.
- [x] Support for serving models that don't have `python-env.yaml`.
- [x] Clean up the environment if `pip install` fails.
- [x] Log `python-env.yaml` in all flavors.
- [x] Restore a model environment using pyenv + virtualenv.
- [ ] ~Update documentation.~ -> Follow up.

## Follow-up tasks

- Replace `--no-conda` with `--env-manager` in tests.
- Add pyenv + virtualenv support for MLflow Projects.

## How is this patch tested?

- Unit tests
- Manual tests using the script below

```python
import mlflow
import subprocess

from sklearn.linear_model import LogisticRegression
from sklearn.datasets import load_iris
from click.testing import CliRunner


X, y = load_iris(return_X_y=True, as_frame=True)
model = LogisticRegression().fit(X, y)

with mlflow.start_run():
    model_info = mlflow.sklearn.log_model(model, "model")

sample_path = "sample.csv"
X.sample(frac=0.1, random_state=0).to_csv(sample_path, index=False)


def run(args):
    args = ["mlflow", "models", *args]
    print("========== Running", args, "==========")
    subprocess.run(args, check=True)


common_args = ["--model-uri", model_info.model_uri, "--env-manager", "virtualenv"]
run(["prepare-env", *common_args])
run(["predict", *common_args, "--input-path", sample_path, "--content-type", "csv"])
run(["serve", *common_args])

# Wait until the server is up and running, and then run this command in another terminal
r"""
curl -X POST -d "{\"data\":[[1,2,3,4]]}" \
-H "Content-Type: application/json; format=pandas-split" \
http://localhost:5000/invocations
"""
```

## Does this PR change the documentation?

- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

## Release Notes

### Is this a user-facing change?


(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 

Interface 

Language 

Integrations


<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:
